### ISSUE TITLE
_TrivialClosureConvert mis-specifies dynamic attributes

### ISSUE THREAD

**Issue by jakevdp** [patrick-kidger/equinox#768]
I believe this should say `static=False`: https://github.com/patrick-kidger/equinox/blob/5a5bf287f73032201c34ff2c73e73573f8232739/equinox/_ad.py#L560

**Comment by patrick-kidger** [comment]
Nope! This is intentional. Given some pytree-of-arrays, then we call`eval_shape` to transform it into a pytree-of-shapedtypestructs which we compare against later, to be sure we are later called with compatible arguments:

https://github.com/patrick-kidger/equinox/blob/5a5bf287f73032201c34ff2c73e73573f8232739/equinox/_ad.py#L546-L549

In this case we actually store the pytree-of-shapedtypestructs as a static `(tuple of shapedtypestructs, treedef)`. This allows the user to pass this object around safely -- across JIT boundaries, `eval_shape` boundaries etc. -- without those `shapedtypestruct`s erroneously getting turned back into dynamic objects.

I appreciate the heads-up though -- I can see how the 'dynamic' in the name suggests otherwise!

**Comment by jakevdp** [comment]
Ah, my mistake, I see what you mean. I think I misdiagnosed this, but there's some real issue related to tracer hashing related to `ClosureConvert`. Here's a minimal repro via diffrax:
```python
import jax
import equinox
import diffrax

print(f"{jax.__version__=}\n{equinox.__version__=}\n{diffrax.__version__=}")
# jax.__version__='0.4.30'
# equinox.__version__='0.11.4'
# diffrax.__version__='0.5.1'
```
```python
# Accelerate the deprecation of tracer hashability, so it raises a TypeError rather than a warning:
from jax._src import deprecations
deprecations.accelerate('tracer-hash')
jax.jit(hash)(1)
# TypeError: unhashable type: <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>
```
```python
from diffrax import diffeqsolve, Dopri5, ODETerm, SaveAt, PIDController

vector_field = lambda t, y, args: -y
term = ODETerm(vector_field)
solver = Dopri5()
saveat = SaveAt(ts=[0., 1., 2., 3.])
stepsize_controller = PIDController(rtol=1e-5, atol=1e-5)

sol = diffeqsolve(term, solver, t0=0, t1=3, dt0=0.1, y0=1, saveat=saveat,
                  stepsize_controller=stepsize_controller)
# TypeError: unhashable type: <class 'jax._src.interpreters.partial_eval.DynamicJaxprTracer'>
```
Something in this call stack ends up computing the hash of a tracer... I thought it may have been in `_TrivialClosureConvert`, but it looks like it's something else (but closely related). Any thoughts on this?

**Comment by patrick-kidger** [comment]
Ah, I see! Haha, looks like we were hashing tracers all along and never even noticed. This should be fixed in #769.

---

For the record -- i.e. future me, if I ever need this -- the problem was that we generate a `_ClosureConvert` on this line:

https://github.com/patrick-kidger/equinox/blob/5a5bf287f73032201c34ff2c73e73573f8232739/equinox/internal/_loop/checkpointed.py#L245

which is a callable that includes some tracers in its pytree structure. This ultimately hits a `lax.while_loop(cond_fun, ...)`. Now `lax.while_loop` hashes its function arguments to try and record their jaxprs, as in https://github.com/google/jax/issues/13554. This previously worked because tracers were hashable and we never even noticed, but when tracers become unhashable then it explodes.

The fix is just to add a dummy `lambda val: cond_fun_(val)` wrapper.
